### ISSUE TITLE
Fatal error: Wrong order of events in the event stream

### ISSUE THREAD

**Issue by vmsgsp** [EventStore/EventStore#858]
When publishing many events (example 20000) with AppendToStreamAsync the order of events is corrupt. The first event appearing in the event stream is not the one having been published first, but the event Nr. 18653.

1. Start the event store cluster node..
EventStore-OSS-Win-v3.4.0\EventStore.ClusterNode --db E:\EventStoreDB

2. Then run the following program:

```
 public class Program
    {
        public static void Main(string[] args)
        {
            var settings = ConnectionSettings.Create();
            settings.LimitConcurrentOperationsTo(1);
            var connection = EventStoreConnection.Create(settings, 
                new IPEndPoint(IPAddress.Parse("127.0.0.1"), 1113));

            connection.ConnectAsync().Wait();

            for (int i = 0; i < 20000; i++)
            {
                connection.AppendToStreamAsync("TestStream", 
                    ExpectedVersion.Any,
                    new EventData(Guid.NewGuid(), "Type" + i, false, new byte[] { }, new byte[] { }));
            }
            while (true) { }
        }
    }
```
The event store stream browser shows the event Type18635 (corresponding to event nr 18635) as first event. See screenshot:
![screenshot](https://cloud.githubusercontent.com/assets/16589636/13915305/2f2ac79a-ef54-11e5-89de-127b962af8dc.jpg)

This is also reproducible when you use the same event type for all events, but write the event number in the data in order to check if the events are published in the right order.

Even if it might not be a real use case to write that many events to the same stream within a very short time period, it worries me that the event order gets corrupted.
A Sleep(1) after each AppendToStreamAsync already solves the problem, but I think this would be a strange workaround.

**Comment by gregoryyoung** [comment]
You just fired off thousands of async tasks there is no assurance which
async tasks finish first

Your sleep helps with ordering the requests.

Try using .Wait() (to wait for the operation to finish before doing the
next one)

On Mon, Mar 21, 2016 at 11:57 AM, vmsgsp <notifications@github.com> wrote:

> When publishing many events (example 20000) with AppendToStreamAsync the
> order of events is corrupt. The first event appearing in the event stream
> is not the one having been published first, but the event Nr. 18653.
>
>    1.
>
>    Start the event store cluster node..
>    EventStore-OSS-Win-v3.4.0\EventStore.ClusterNode --db E:\EventStoreDB
>    2.
>
>    Then run the following program:
>    public class Program
>    {
>    public static void Main(string[] args)
>    {
>    var settings = ConnectionSettings.Create();
>    settings.LimitConcurrentOperationsTo(1);
>    var connection = EventStoreConnection.Create(settings,
>    new IPEndPoint(IPAddress.Parse("127.0.0.1"), 1113));
>
>          connection.ConnectAsync().Wait();
>
>          for (int i = 0; i < 20000; i++)
>          {
>              connection.AppendToStreamAsync("TestStream",
>                  ExpectedVersion.Any,
>                  new EventData(Guid.NewGuid(), "Type" + i, false, new byte[] { }, new byte[] { }));
>          }
>          while (true) { }
>      }
>
>    }
>
> The event store stream browser shows the event Type18635 (corresponding to
> event nr 18635) as first event. See attached screenshot.
> [image: screenshot]
> <https://cloud.githubusercontent.com/assets/16589636/13915106/eeed5676-ef52-11e5-8fec-53d889b53d85.jpg>
>
> This is also reproducible when you use the same event type for all events,
> but write the event number in the data in order to check if the events are
> published in the right order.
>
> Even if it might not be a real use case to write that many events to the
> same stream within a very short time period, it worries me that the event
> order gets corrupted.
> A Sleep(1) after each AppendToStreamAsync already solves the problem, but
> I think this would be a strange workaround.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/EventStore/EventStore/issues/858>
>



-- 
Studying for the Turing test

**Comment by vmsgsp** [comment]
I wanted to publish events in a non-blocking way and use the KeepReconnecting feature of the event store in case the event store process is down for (short) while. Of course this would not work with too many events because the client side buffer (implemented in EventStore.ClientAPI) has some natural limits. But I would expect that when I call AppendToStreamAsync the EventStore.ClientAPI would execute a very small piece of synchronous code to ensure the right event ordering before creating the asynchronous task.

**Comment by gregoryyoung** [comment]
I'm just curious how exactly would we maintain ordering for you when you
ran into say a retry?

On Mon, Mar 21, 2016 at 12:12 PM, vmsgsp <notifications@github.com> wrote:

> I wanted to publish events in a non-blocking way and use the
> KeepReconnecting feature of the event store in case the event store process
> is down for (short) while. Of course this would not work with too many
> events because the client side buffer (implemented in EventStore.ClientAPI)
> has some natural limits. But I would expect that when I call
> AppendToStreamAsync the EventStore.ClientAPI would execute a very small
> piece of synchronous code to ensure the right event ordering before
> creating the asynchronous task.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/EventStore/EventStore/issues/858#issuecomment-199210654>
>



-- 
Studying for the Turing test

**Comment by vmsgsp** [comment]
Publish the events in the same order as you have queued them in the client side buffer. Concurrency between separate processes of course cannot be solved through this, but in our environment this is solved by other means (not by the event store).

**Comment by gregoryyoung** [comment]
Let's try a concrete example. Its basically how it works. On async put operation in queue.

There is a background process picking up those operations off the queue and sending them over TCP (note the protocol is fully asynchronous). So send 1,2,3,4,5,6,7,8,9 receive ack 1, ack 2, ack 3, timeout 4, ack 5 what exactly should we do with 4 in order to get your ordering for you?

**Comment by vmsgsp** [comment]
One solution could be: "Do not send 5 before 4 has been acknowledged.". If there is a timeout in TCP the scenario is similar to "Event Store service down => reconnect when up again".

In the end it seems to be a question of in which reliability mode the event store shall be used:
Mode 1: Do not loose any events => better try to store all you can, no matter in which order
Mode 2: Never change the order of events => accept the higher risk of loosing some events at the end of the stream

**Comment by gregoryyoung** [comment]
"One solution could be: "Do not send 5 before 4 has been acknowledged.". If there is a timeout in TCP the scenario is similar to "Event Store service down => reconnect when up again"."

This is exactly what .Wait() / await would do! And I wasn't saying a TCP timeout as that would have prevented message 5 from arriving. 

When flooding async operations the connection makes a best effort to keep ordering but it is not assured (whether it sheds messages vs retries it etc is set via configuration)

**Comment by vmsgsp** [comment]
I have seen a re-ording of events for only 3 events as well and then came to the solution of using LimitConcurrentOperationsTo(1) (see https://groups.google.com/forum/#!topic/event-store/Ox2O-_MV-78). And I thought this would solve it for more than 3 events as well.

If I go with the the Wait() solution I cannot use the reconnection feature of even store any more (without blocking my publishing application). In the end I would have an own buffer of tasks being waiting...

**Comment by gregoryyoung** [comment]
Limitconcurrentoperations(1) says only allow a single concurrent operation over the tcp socket this is not something you would generally want to use (your connection will be idle 90% of the time)

"If I go with the the Wait() solution I cannot use the reconnection feature of even store any more (without blocking my publishing application). In the end I would have an own buffer of tasks being waiting..."

I really don't understand why you would need to do all this work. As example.

void Publish(Message m) {
   _queue.Enqueue(m);
}

public void ReadsFromQueue() {
    while(true) {
         Message m;
         if(_queue.TryDequeue(out m)) {
               //appendasync.wait()
         }
    }
}

new Task(() => ReadsFromQueue(),
                    TaskCreationOptions.LongRunning);

Given I would probably not loop like this but...

**Comment by gregoryyoung** [comment]
I am not understanding perhaps the use case of "we need perfect ordering but we don't care if writes succeed". Maybe you can explain more about the use case?

I will look at the one write that came up first as this is likely something that shouldn't be reordered, most of this is more general in that there are edge cases where you will not receive perfect ordering of async operations if just using fire/forget.

**Comment by vmsgsp** [comment]
The stream of events we are writing contains the change events of a single aggregate instance. So the state of the aggregate can only be re-established when the change events are replayed in the correct order. We do not care about the order of events across aggregate instances, but shouldn't the order be maintained per aggregate instance?
Actually we care if the write succeeds, but we assume to have a retry-mechanism in place also in case the sending process crashes.

**Comment by gregoryyoung** [comment]
"Actually we care if the write succeeds, but we assume to have a retry-mechanism in place also in case the sending process crashes."

Shutting down/crashing your server with your writes cached in memory?
Getting an error from the event store?

"So the state of the aggregate can only be re-established when the change events are replayed in the correct order"

The events that are still somewhere between you asynchronously writing them and your read?

**Comment by vmsgsp** [comment]
The real truth of our aggregate is always persisted in a SQL database where every save operation is blocking (in terms of: the user will be notified IMMEDIATELY if saving fails) and fully transactional.

We use the event store in order to improve the user-perceived responsiveness and performance of our system. That is: when an aggregate changes it fires events, based on which some heavy-weight background calculation services can be started for other aggregates that are dependent on the changed aggregate. And for this we also need the correct order of events. The background processing shall ensure that all data is prepared and ready once the user is accessing the affected aggregates the next time.

In the worst case of lost events or corrupted event order we will detect this (through our SQL database) and can re-run all calculations needed (but then the user has to wait a little bit). So in terms of data consistency and correctness we are not dependant on a 100% reliability of the event storing and processing, but the goal is to be as reliable as possible.

Our reasons to use the event store:
1. We use it as message bus for inter-process event communication
2. System-runtime-decoupling: Storing events locally that have been published by another remote system, allows to run the background event processing (locally) even while the remote system is being unreachable or not running
3. We like the functionality of catch-up subscriptions and the reconnecting feature of the event store, and we know how complex it would be to implement this by ourselves (against our SQL database).

Finally, we could implement the message queue as you have outlined above in order to ensure correct event ordering, but as said, this is yet another additional queue to the queue(s) we already have in the event store client API. And this looks like smearing / distributing core functionality of event publishing over the event store client API and our own custom implementation.

You said:
"When flooding async operations the connection makes a best effort to keep ordering but it is not assured (whether it sheds messages vs retries it etc is set via configuration)":
-"Best effort" is acceptable for us (as described above we do not need 100% reliability), but I really tried with 3 events 1-2-3 only and they are published in the order 2-1-3, which does not look like best effort.
-Which configuration parameters do you mean?

**Comment by riccardone** [comment]
"The stream of events we are writing contains the change events of a single aggregate instance."

Hi, 
Given that the AppendToStreamAsync connection method can accept multiple events and write them atomically, why you don't send the events related to each aggregate in one operation instead of splitting them in a flow of separate events?

**Comment by vmsgsp** [comment]
I can have several transactions on my aggregate (let's say within 5 minutes) but the event store is down during this time. So I would have to write my own order-keeping buffer again.

**Comment by riccardone** [comment]
It seems to me that you are writing "infrastructure code". 
I rather consider EventStore like a piece of infrastructure for your Domain. If you need EventStore to be high available you can do that. In any case, if the EventStore is down after connection retries my Domain host service kill automatically itself and I set a service monitor that notify Ops.

**Comment by vmsgsp** [comment]
This is how we are trying to use event store: as a proven piece of infrastructure. Back to the event stream of an aggregate instance: Isn't it a strong logical property that the event stream is ordered? And shouldn't the infrastructure maintain this property also over reconnects?

**Comment by riccardone** [comment]
Yes, the EventStore is the infrastructure and the client connection is a proxy to it with retry feature. ES can receive and save multiple Aggregate events atomically. You can send from your client in that way so that the stream of events are ordered. In case of disaster and ES goes down I prefer stop the client service and fix the infrastructure instead of expecting or implementing a local queue.

**Comment by vmsgsp** [comment]
So if I understand you correctly, you do not use the "reconnect"-feature of ES?

**Comment by gregoryyoung** [comment]
For original question I think what you are seeing is a difference in

https://github.com/EventStore/EventStore/blob/release-v3.6.0/src/EventStore.ClientAPI/Internal/EventStoreConnectionLogicHandler.cs#L376

Which then calls Schedule/Enqueue

https://github.com/EventStore/EventStore/blob/release-v3.6.0/src/EventStore.ClientAPI/Internal/OperationsManager.cs#L189

https://github.com/EventStore/EventStore/blob/release-v3.6.0/src/EventStore.ClientAPI/Internal/OperationsManager.cs#L183

If you enable verbose logging for your test this can be confirmed.

**Comment by gregoryyoung** [comment]
.EnabledVerboseLogging().UseCustomLogger() 
.EnabledVerboseLogging().UseConsoleLogger()
.EnabledVerboseLogging().UseFileLogger()
etc

**Comment by vmsgsp** [comment]
I add this to my publishing client? Where do I see the logfile? What should I see in the logfile?

**Comment by gregoryyoung** [comment]
The FileLogger takes a string which is the filename.

The log file will contain verbose logging from the client api.

**Comment by gregoryyoung** [comment]
btw I am pretty sure I have isolated (and have a fix) for the one message above. The logs will allow me to verify what is happening, providing it still happens with logs there is a good chance the behaviour will change with logging enabled. The fix while fixing the above issue will still not give assurances of perfect ordering in the case of retries etc.

**Comment by vmsgsp** [comment]
Please find the logfile here:
[EventStoreLog.zip](https://github.com/EventStore/EventStore/files/184050/EventStoreLog.zip)
BTW: I deleted my event store by accident, and then I downloaded version 3.5.0 (previously I had 3.4.0). But the wrong ordering is still reproducible. Event 1638 appears first in my stream.

**Comment by gregoryyoung** [comment]
This appears truncated (did you copy/paste off console etc?) the first operation in it is:

ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': HandleTcpPackage connId a6ea0a5c-3a9d-46f7-b79e-dd131822826f, package WriteEventsCompleted, 520f41c7-017f-4426-bc7f-242c4b916cf2..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': HandleTcpPackage OPERATION DECISION EndOperation (Success), Operation AppendToStreamOperation

**Comment by vmsgsp** [comment]
I took the entire file and zipped it. But it starts with StartConnectionMessage. What do you expect in front of this? What else could I try?

**Comment by vmsgsp** [comment]
When I download the attached file again, I also see the following lines at the beginning. Maybe you are using a different zip-tool than I did (I used Windows "Send to zip file" function).

Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.StartConnectionMessage..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': StartConnection.
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': DiscoverEndPoint.
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.EstablishTcpConnectionMessage..

**Comment by gregoryyoung** [comment]
Ah I opened it with logviewer on mac which truncated it.

Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.StartOperationMessage..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.StartOperationMessage..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': ScheduleOperation package WriteEvents, fada0729-a620-4b54-b247-b8b68670df96, Operation AppendToStreamOperation (fada0729-a620-4b54-b247-b8b68670df96): Stream: TestEventOrder, ExpectedVersion: -2, retry count: 0, created: 12:06:19.715, last updated: 12:06:19.716..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.StartOperationMessage..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.StartOperationMessage..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message EventStore.ClientAPI.Internal.StartOperationMessage..
Debug: EventStoreConnection 'ES-d2c91474-c4b8-4292-9b39-b80b5a43aedd': enqueueing message 

Is exactly what I expected to see

The issue is right here. https://github.com/EventStore/EventStore/blob/release-v3.6.0/src/EventStore.ClientAPI/Internal/OperationsManager.cs#L193 there could be queued items here.

I will pull request likely this afternoon or tomorrow (depending what interrupts my day) and it should resolve this case. I want to reiterate though that there are other cases where its fully expected you will get out of order message processing.

**Comment by vmsgsp** [comment]
**First**: Thank you very much for following up on this.

**Second**: Regarding your statement
"I want to reiterate though that there are other cases where its fully expected you will get out of order message processing.":
- Does this mean that even for my single aggregate instance writing the change events into one dedicated stream, there is a potential of reading events in the wrong order?

**Comment by gregoryyoung** [comment]
"Does this mean that even for my single aggregate instance writing the change events into one dedicated stream, there is a potential of reading events in the wrong order?"

Yes for examples given above on some edge conditions like reconnects/errors/timeouts etc. Limiting concurrent operations to 1 and setting keepretrying gets rid of many of these issues but not all. It also introduces a whole new set of issues around management (what happens if you consistently get a timeout on one message, should it retry forever never processing any others?). There are also issues of performance with concurrentoperations =1 especially talking to a cluster as your performance will be essentially blocking single threaded.

**Comment by gregoryyoung** [comment]
Been trying to reproduce this but can't on my machine (which is not surprising since its a threading issue)

With the patch I get:

![image](https://cloud.githubusercontent.com/assets/381274/13955164/933ffb32-f03b-11e5-81e0-15cb63d5e47c.png)


My testing setup is:

using System;
using System.Net;
using EventStore.ClientAPI;
namespace test {
    public class Program
    {
        public static void Main(string[] args)
        {
            var settings = ConnectionSettings.Create();
            settings.LimitConcurrentOperationsTo(1);
            var connection = EventStoreConnection.Create(settings,
                new IPEndPoint(IPAddress.Parse("127.0.0.1"), 1113));

            connection.ConnectAsync().Wait();

            for (int i = 0; i < 20000; i++)
            {
                connection.AppendToStreamAsync("TestStream",
                    ExpectedVersion.Any,
                    new EventData(Guid.NewGuid(), "Type" + i, false, new byte[] { }, new byte[] { }));
            }
            while (true) { }
        }
    }
}

built/ran in mono.


You can try it here: https://github.com/EventStore/EventStore/tree/client-ordering build and client api is in bin/clientapi

**Comment by vmsgsp** [comment]
Witht the client-ordering build I get a StackOverflowException (infinite recursive loop):
>	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 175	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 200	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 178	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 200	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 178	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 200	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 178	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 200	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 178	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 200	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 178	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 200	C#
 	EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection connection) Line 178	C#

**Comment by gregoryyoung** [comment]
Try again there are two commits there

On Tue, Mar 22, 2016 at 3:22 PM, vmsgsp <notifications@github.com> wrote:

> Witht the client-ordering build I get a StackOverflowException (infinite
> recursive loop):
>
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 175 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem
> operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 200 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 178 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem
> operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 200 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 178 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem
> operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 200 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 178 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem
> operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 200 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 178 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem
> operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 200 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 178 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.ScheduleOperation(EventStore.ClientAPI.Internal.OperationItem
> operation, EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 200 C#
> EventStore.ClientAPI.dll!EventStore.ClientAPI.Internal.OperationsManager.TryScheduleWaitingOperations(EventStore.ClientAPI.Transport.Tcp.TcpPackageConnection
> connection) Line 178 C#
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/EventStore/EventStore/issues/858#issuecomment-199862869>
>



-- 
Studying for the Turing test

**Comment by vmsgsp** [comment]
Nice, with the latest version at least I cannot reproduce the issue anymore (which is not a proof, but a good first indication).
It also works with publishing 20000 events before starting the even store (using reconnecting). No events lost and order correct also with this reconnect scenario. 

You said
"It also introduces a whole new set of issues around management (what happens if you consistently get a timeout on one message, should it retry forever never processing any others?). "

Well, good questions. To be honest I did not think that deep into these challenges. I had a (maybe too) naïve view on the capabilities of the event store. I thought "it just does all this for me".

So what has the KeepReconnecting been developed for? Only for subscriptions? Why does it "apparently" work also for publishing? Maybe a little more documentation could help a lot here...

**Comment by gregoryyoung** [comment]
"Well, good questions. To be honest I did not think that deep into these challenges. I had a (maybe too) naïve view on the capabilities of the event store. I thought "it just does all this for me".

So what has the KeepReconnecting been developed for? Only for subscriptions? Why does it "apparently" work also for publishing? Maybe a little more documentation could help a lot here..."

KeepReconnecting will work great for publishing messages as well. It handles reconnecting the connection. KeepRetrying works great too as does limiting the number of concurrent requests. The problem can come in when you start saying things like KeepReconnecting().KeepRetrying().LimitConcurrentOperations(1) in this case if you end up with a message that times out deterministically for some reason you will get "stuck". This is not due to the APIs not working, they are working exactly as you told them to. Timeout->retry->timeout->retry->timeout-retry and you have limited concurrent operations to 1 which means it will sit in this loop forever. At some point you probably want it to stop retrying (especially if concurrent limit is 1). Once you do this you run the risk of a write not completing due to retry timeouts, which is likely something you want to know about.

**Comment by gregoryyoung** [comment]
As an update if you set concurrentoperations higher than 1 you will as of now receive out of order writes. This is due to having multiple readers that are being used for things like verifying security. @pgermishuys  has pushed up a PR that gives affinity to these messages based on stream (so they will still be ordered within streams). We are discussing adding a flag to give global affinity (at the possible cost of performance) after setting which you should see ordering even with .LimitConcurrentOperations(100) with the possible exception of some of the caveats listed above

**Comment by gregoryyoung** [comment]
See https://github.com/EventStore/EventStore/pull/862 it also includes the better ordering queue affinity option.

I will close this issue for now.

Cheers,

Greg

**Comment by vmsgsp** [comment]
" @pgermishuys has pushed up a PR that gives affinity to these messages based on stream (so they will still be ordered within streams)."
=> sorry I could not find this PR. Could you please send a link?

"As an update if you set concurrentoperations higher than 1 you will as of now receive out of order writes"
=> Why "as of now". I think this behaved like this before your changes as well?

**Comment by pgermishuys** [comment]
This is the PR in question.
https://github.com/EventStore/EventStore/pull/862
Both @gregoryyoung and I have commits in there.

**Comment by gregoryyoung** [comment]
"=> Why "as of now". I think this behaved like this before your changes as well?" i refering to now (before the changes)
